src/main/java/org/openpnp/machine/reference/vision/ReferenceBottomVision.java

```diff
@@ -65,10 +68,77 @@ public PartAlignmentOffset findOffsets(Part part, BoardLocation boardLocation,
+			part=nozzle.getPart();
+			double angle = placementLocation.getRotation();
+			if(boardLocation!=null) {
+				angle = Utils2D.calculateBoardPlacementLocation(boardLocation, placementLocation).getRotation();              
+			}
+			while (Math.abs(angle) > 180.0) { angle += (angle < 0.)? 360. : -360.; } 
+			double nozzleAngle=angle;
+			double placementAngle=angle;
+            MovableUtils.moveToLocationAtSafeZ(nozzle, VisionUtils.getBottomVisionCamera().getLocation().add(new Location(part.getHeight().getUnits(), 0.0, 0.0,part.getHeight().getValue(), 0.0)).derive(null, null, null, angle));
+            CvPipeline pipeline = partSettings.getPipeline();
+            pipeline.setCamera(VisionUtils.getBottomVisionCamera());
+            pipeline.setNozzle(nozzle);
+			pipeline.setValue(angle);
+			if(Boolean.getBoolean("develop")) {
+            if ("Rotate".equals(pipeline.getStages().get(1).getClass().getSimpleName())) {
+                ((Rotate)pipeline.getStages().get(1)).setDegrees( angle );
+            }}
+			pipeline.process();
+			if (!((pipeline.getResult("result")).model instanceof RotatedRect)) {
+				throw new Exception("Bottom vision alignment failed for part " + part.getId()
+						+ " on nozzle " + nozzle.getName() + ". No result found.");
+			}
+
+
+			RotatedRect rect = ((RotatedRect)(pipeline.getResult("result")).model);
+			if (rect.size.width < rect.size.height) { rect.angle = 90 + rect.angle; }
+            Logger.debug("Result rect {}", rect);
+			while (Math.abs(angle) > 45.0) { nozzleAngle= (angle += (angle < 0.)? 90 : -90); } 
+			while (Math.abs(rect.angle) > 45.0) { rect.angle += (rect.angle < 0)? 90 : -90; } 
+			angle += rect.angle;
+			while (Math.abs(angle) > 45.0) { angle += (angle < 0.)? 90 : -90; } 
+
+			int cnt=1;
+			if (pipeline.getResult("contours")!=null&&pipeline.getResult("contours").model instanceof List<?>) {
+				cnt=((List<?>)(pipeline.getResult("contours")).model).size();
+			}
+
+			String s="";
+			Location offsets = null;
+			int old=999;
+			int rot=999; do {	
+				offsets = new Location(LengthUnit.Millimeters,Double.NaN, Double.NaN, Double.NaN, placementAngle+angle);
+				nozzle.moveTo(offsets, part.getSpeed());
+				pipeline.process();
+				if (!((pipeline.getResult("result")).model instanceof RotatedRect)) {
+					throw new Exception("Bottom vision alignment failed for part " + part.getId()
+							+ " on nozzle " + nozzle.getName() + ". No result found.");
+				}
+
+				rect = (RotatedRect) pipeline.getResult("result").model;
+		                Logger.debug("Result rect {} inside vision loop", rect);
+				offsets = VisionUtils.getPixelCenterOffsets(pipeline.getCamera(), rect.center.x, rect.center.y).derive(null,null,null,-angle);
+				s = "Align offset for "+part.getName()+": " + offsets.toString() + "     " ;
+				MainFrame.get().setStatus(s);
+				offsets = offsets.derive(null,null,null,Double.NaN);
+				if(cnt>Integer.getInteger("contours",40)||Double.isNaN(pipeline.getValue())) {
```

Comment:
This is a feature for better usage.
If instead of MinAreaRect  the follow stages get used:
Threshold with value of min from MinAreaRect
findContours  external
MinAreaRectContours with findContours stage
The result is the same, but findContours result hold the number of discovered pins.
It has to be noted that each pin on package like soic/tssop/qfp/... each pin get detected with
two contours, and 32 pins get 64 contours on correct illumination. Because not all have correct illumination, maybe 80 or 70% of pins get correctly detected, and the number is then near 50
for 32pin tqfp package.  Now if vision detect that part have more then 40 pins, it uses automatically closed loop instead of prerotate. To enable this behaviour, the findContour stage
mentioned above must be named "contours" and with -Dcontours=20 that value could be lowered or increased. For tqfp32 it is never needed, but for ram as bga on wson package as example yes.  It was a user request during the extended beta test.